File: backend/src/utils/embedding/adapters/ollama.ts
import type { EmbeddingAdapter, EmbeddingProviderRecord } from '../types';
import { EmbeddingError } from '../types'
import { DEFAULT_MODELS, parseOpenAIResponse } from './base'

// Compat mode: endpoint contains /v1 — uses OpenAI-compatible /v1/embeddings
// Native mode: bare Ollama endpoint — uses /api/embed (supports dimensions)
function isCompatMode(provider: EmbeddingProviderRecord): boolean {
  return provider.endpoint.includes('/v1')
}

export class OllamaAdapter implements EmbeddingAdapter {
  getEndpointPath(provider: EmbeddingProviderRecord): string {
    return isCompatMode(provider) ? '/embeddings' : '/api/embed'
  }

  buildHeaders(_provider: EmbeddingProviderRecord): Record<string, string> {
    return { 'Content-Type': 'application/json' }
  }

  buildBody(provider: EmbeddingProviderRecord, input: string): Record<string, unknown> {
    const model = provider.model || DEFAULT_MODELS['ollama']
    // Both compat and native modes use 'input' key.
    // /api/embed and /v1/embeddings both accept 'dimensions' for MRL models.
    const body: Record<string, unknown> = { input }
    if (model) body['model'] = model
    if (provider.dimensions !== null) body['dimensions'] = provider.dimensions
    return body
  }

  parseResponse(data: unknown): number[] {
    // OpenAI compat response: { data: [{ embedding: [...] }] }
    if ((data as { data?: unknown })?.data !== undefined) {
      return parseOpenAIResponse(data)
    }
    // Ollama native /api/embed response: { embeddings: [[...]] }
    const d = data as { embeddings?: unknown[] }
    if (Array.isArray(d?.embeddings) && d.embeddings.length > 0) {
      const first = d.embeddings[0]
      if (Array.isArray(first) && first.length > 0) return first as number[]
    }
    throw new EmbeddingError('Invalid Ollama response: expected embeddings[0] array')
  }
}

File: gateway-engine/src/adapters/embedding/ollama.rs
//! Ollama embedding adapter — two API modes, selected from the endpoint:
//!   - **compat**: endpoint contains `/v1` → OpenAI-compatible `/embeddings`.
//!   - **native**: otherwise → Ollama's own `/api/embed` (`embeddings[0]` shape).

use serde_json::Value;

use crate::adapters::{check_provider_error, is_ollama_compat, json_headers};
use super::{extract_f32, parse_openai_response, standard_body, EmbeddingAdapter};
use crate::agents::embedding::client::EmbeddingProviderConfig;

pub struct OllamaEmbeddingAdapter {
    /// True when the endpoint exposes the OpenAI-compatible `/v1` surface.
    compat: bool,
}

impl OllamaEmbeddingAdapter {
    pub fn new(endpoint: &str) -> Self {
        OllamaEmbeddingAdapter { compat: is_ollama_compat(endpoint) }
    }
}

impl EmbeddingAdapter for OllamaEmbeddingAdapter {
    fn vendor(&self) -> &'static str {
        "ollama"
    }

    fn endpoint_path(&self, _p: &EmbeddingProviderConfig) -> String {
        if self.compat { "/embeddings".to_string() } else { "/api/embed".to_string() }
    }

    fn build_headers(&self, _p: &EmbeddingProviderConfig) -> Vec<(String, String)> {
        let hm = json_headers();
        (&hm).into_iter().map(|(k, v)| (k.to_string(), v.to_str().unwrap().to_string())).collect()
    }

    fn build_body(&self, p: &EmbeddingProviderConfig, input: &str) -> Value {
        standard_body(p, input)
    }

    fn parse_response(&self, data: &Value) -> Result<Vec<f32>, String> {
        if let Some(msg) = check_provider_error(data) {
            return Err(format!("provider error: {}", msg));
        }
        if self.compat {
            return parse_openai_response(data);
        }
        // Native: { "embeddings": [[...]] }
        let arr = data["embeddings"][0].as_array()
            .ok_or_else(|| "ollama response missing \"embeddings[0]\" array".to_string())?;
        extract_f32(arr, "embeddings[0]")
    }
}
